hail/src/main/scala/is/hail/types/virtual/TIterable.scala

```diff
@@ -1,10 +1,21 @@
 package is.hail.types.virtual
 
-import is.hail.types.physical.PIterable
+import is.hail.types.BaseType
 import is.hail.utils.FastSeq
 
 abstract class TIterable extends Type {
   def elementType: Type
 
   override def children = FastSeq(elementType)
 }
+
+object TIterable {
+  def elementType: BaseType => Type = {
```

Comment:
This is a prime example of what I call Scala BS: something the language makes easy to write that seems fine, but when you think and look hard is horrifying. As an exercise for you, investigate why this will have worse performance than the method I've written below (it's not the match):
```
  def elementType(t: Type): Type = t.asInstanceOf[TIterable].elementType
```

Reply:
Ok, though I really don't think we should be worrying about making micro-optimisations like this or a few extra allocs. If it's a problem and a profiler says it's a hotspot, then we should care, but not before.

Reply:
For posterity, using a lambda generates an additional load and `invokedynamic`: https://godbolt.org/z/sx6TfKbeG